### ISSUE TITLE
PartialOrd for Level and PartialOrd<LevelFilter> for Level is implemented in the wrong direction

### ISSUE THREAD

**Issue by juumixx** [tokio-rs/tracing#2367]
Right now, `Level.ERROR < Level.TRACE`, even though `LevelInner::Error = 4` and `LevelInner::Trace = 0`. This is very confusing and was already noticed in various other issues regarding documentation.

Here is the culprit:
https://github.com/tokio-rs/tracing/blob/68f830d392b3b2fc07704d107d223414a1291da4/tracing-core/src/metadata.rs#L920
and
https://github.com/tokio-rs/tracing/blob/68f830d392b3b2fc07704d107d223414a1291da4/tracing-core/src/metadata.rs#L949

The comparison functions should be implemented the other way round, like this:
```
    fn lt(&self, other: &Level) -> bool {
        (self.0 as usize) < (other.0 as usize)
    }
```

This would invert current comparison logic, but in the end it would behave as expected (by many). Existing documentation would also be affected (see https://github.com/tokio-rs/tracing/issues/1669)

**Comment by davidbarsky** [comment]
This is not a bug. This happens for two reasons:
- `Level::ERROR` should be considered _less than_ `Level::TRACE` because `Level::ERROR` is _less verbose_ that `Level::TRACE`. In other words, `Level::TRACE` permits more things to be emitted.
- The comparison, due to `LevelInner`, needs to be inverted for performance optimization reasons. This is explained in this (private, granted) comment here: https://github.com/tokio-rs/tracing/blob/68f830d392b3b2fc07704d107d223414a1291da4/tracing-core/src/metadata.rs#L873-L900

**Comment by juumixx** [comment]
Thanks for the explanation. When comparing the verbosity, it makes sense.
Most frameworks I know compare the criticality, not the verbosity of a log level, though, so it feels really unintuitive to me.
Even after reading the source with `ERROR` having a higher ordinal than `TRACE` it felt like it should be the other way round. Maybe it is enough to move the documentation closer to the actual implementation? I must admit, I didn't read this at all because the `BIG, IMPORTANT WARNING` was way out of my scope when I was looking for the reason for this behavior.

I cannot judge how useful this optimization is, it just makes me sad when performance has to be chosen over ergonomics :(

**Comment by hawkw** [comment]
The performance optimization discussed in the comment that @davidbarsky linked to does not impact the user-facing comparison behavior, only the internal representation of the `Level` types. Performance concerns here influence only how `Level`s are represented in memory, not user-facing behavior.

Instead, the decision that `ERROR` should compare as the "lowest" level and `TRACE` as the "highest" was made long before the optimization in the comment @davidbarsky linked to was added. We decided on that ordering because it's [the order used by the `log` crate, where `Level::Error`](https://docs.rs/log/latest/log/enum.Level.html) will compare as the "lowest" level. Since the `log` crate had very widespread adoption in the Rust ecosystem when we started implementing `tracing`, we used the same level ordering behavior (and names) because we thought a majority of potential `tracing` users would be familiar with `log`. While libraries in other programming languages may do things differently, we've prioritized consistency with the rest of the Rust library ecosystem.
